### DISCUSSION TITLE
How to use filterFields for pivot model

### DISCUSSION THREAD

**Discussion by goldmont** [wintercms/winter#1058] [General]
### Winter CMS Build

dev-develop

### PHP Version

8.1

### Database engine

MySQL/MariaDB

### Plugins installed

None

### Issue description

Hi there.

I have the following scenario:
I have **many** `Devices` that could be related to **many** `Dashboards`. The `pivot` table has also additional attributes set via a `pivot` form. There is a `pivotModel` that extends `Winter\Storm\Database\Pivot`. From the `Dashboard` form, I can link many devices and, every time I choose a device, the `pivot` form is displayed to allow me to enter `pivot` additional attributes.

I'm trying to manipulate `pivot` form via `filterFields` method but I noticed that this last not called on `pivotModel` but on `Device::class`. From `Device::class` I'm unable to access `pivot` form fields. However `Accessors` and `Mutators` are correctly called on `pivotModel`.

### Steps to replicate

`Dashboard` model:
```php
// ...

public $belongsToMany = [
  'devices' => [
    Device::class,
    'table' => 'iot_dashboards_devices',
    'key' => 'dashboard_id',
    'otherKey' => 'device_id',
    'pivot' => ['type', 'L1', 'L2', 'L3'],
    'pivotModel' => DashboardsDevicesPivot::class
  ]
];
```

I tried to implement `relationExtendPivotWidget($widget, $field, $model)` method into the `Dashboards` controller and I'm able to manipulate the `pivot` form. By the way `$widget->model->pivot` data is an empty array unless the `context` is `update`.

Like this:
```php
public function relationExtendPivotWidget($widget, $field, $model)
{
  $data = $widget->model->pivot;
  // $data is empty
  $widget->fields['pivot[my_field]']['hidden'] = true;
}
```

### Workaround

I've also noticed that if I call manually the method `$widget->getSaveData()` inside `relationExtendPivotWidget($widget, $field, $model)` I can get `pivot` form data sent via AJAX request.

Like this:
```php
public function relationExtendPivotWidget($widget, $field, $model)
{
  $data = $widget->getSaveData();
  $type = $data['type']; // Getting "type" additional pivot attribute.
  $widget->fields['pivot[my_field]']['hidden'] = true;
}
```

By the way, for some reason, after calling `$widget->getSaveData();`, the line `$widget->fields['pivot[my_field]']['hidden'] = true;` has no effect.

**Comment by LukeTowers** [comment]
@mjauvin would you be able to take a look?

**Comment by mjauvin** [comment]
@goldmont if you could submit a PR to the Winter.Test plugin's User/UserRolePivot models to demonstrate this, that would help move this forward for investigation.

Thanks.

**Comment by mjauvin** [comment]
@goldmont you cannot use the Pivot Model for filterFields, use `filterFields` on the relation Model (Device model in your case)

**Comment by goldmont** [comment]
> @goldmont you cannot use the Pivot Model for filterFields, use `filterFields` on the relation Model (Device model in your case)

Hi,

As you said, I've managed to alter pivot form using `filterFields` on my relation Model (`Device` model).

I had to dynamically access the field using the following syntax:
```php
public function filterFields($fields, $context = null)
{
	$fields->{'pivot[my_field]'}->hidden = true;
}
```

By the way, I'm still unable to get the value of a dropdown `pivot[my_dropdown]` in order to alter multiple `pivot[*]` fields. How can I solve? Thanks.

**Reply by mjauvin** [reply]
Isn't `$fields->{'pivot[dropdown]'}->value` working ?
